### PR TITLE
Mulivariate gaussian family

### PR DESCRIPTION
**Pull request by jolars**
This pull request introduces the multivariate gaussian family to sgdnet.

```r
fit <- sgdnet(student$x, student$y, family = "mgaussian")
plot(fit)
```
image

There are also a rather large portion of changes under the hood. Notably, many of the STL types of vectors have been replaced in favor of versions from Eigen. I believe this makes the code more readable and concise. The code in [saga.h](https://github.com/jolars/sgdnet/blob/mulivariate-gaussian/src/saga.h), for instance, is down to 403 lines from 547.

Another change is that the penalties have been made into functors, now available in <https://github.com/jolars/sgdnet/blob/mulivariate-gaussian/src/penalties.h>. I would appreciate it if you took a look at the implementation of the group lasso here to see that I've got it right.

There is also a new data set to be used in examples and the tests have gotten an overhaul in several places.

The travis.ci integration has been expanded to cover OSX and all releases of R. The appveyor build is also activated now and should show up with this pull request.

### CODE REVIEW

**Comment by michaelweylandt** [comment]
This is too much for me to review all at once - I did go through the changes to the R-level tests and they generally look good. 

I *think* the group lasso penalty looks correct, but it might be easier to implement it differently. The math is `(1 - lambda / |beta|_2)_+ * beta`, so I'd just calculate the scalar prefactor (zero-ing if negative) and then multiply by that. I'm not 100% sure how you're doing the scaling, so I'm not sure what it would be in your notation.

File: src/families.h [hunk 1]
```diff
@@ -126,149 +137,232 @@ class Binomial : public Family {
...
+  double
+  Loss(const Eigen::ArrayXd&  linear_predictor,
+       const Eigen::MatrixXd& y,
+       const unsigned         i) const noexcept
+  {
 
-    for (unsigned j = 0; j < n_classes; ++j)
-      y_mu[j] = y_mat.col(j).mean();
+    return 0.5*(linear_predictor - y.array().col(i)).square().sum();
+  }
```
Comment (michaelweylandt) [inline src/families.h]: Is `y` being stored "transposed" internally? 

I would normally interpret the `i`-th column of `y` as being the `i`-th response for all observations, not all responses for the `i`-th observation.

File: src/sgdnet.cpp [hunk 2]
```diff
@@ -131,45 +130,50 @@ Rcpp::List SetupSgdnet(T                   x,
                      beta,
                      family);
 
+  // Transpose for more efficient access of samples
+  AdaptiveTranspose(x);
+  AdaptiveTranspose(y);
```
Comment (michaelweylandt) [inline src/sgdnet.cpp:135]: Ignore other review comment - just saw this. (Or even better - maybe add a comment referring to this transpose in the MultiGaussian family implantation)

File: src/sgdnet.cpp [hunk 3]
```diff
@@ -131,45 +130,50 @@ Rcpp::List SetupSgdnet(T                   x,
...
   // Null deviance on scaled y for computing deviance ratio
-  double null_deviance_scaled = family.NullDeviance(y,
-                                                    fit_intercept,
-                                                    n_classes);
+  double null_deviance_scaled = family.NullDeviance(y, fit_intercept);
+
   vector<double> deviance_ratio;
   deviance_ratio.reserve(n_lambda);
 
+  // Rcpp::stop("here");
```
Comment (michaelweylandt) [inline src/sgdnet.cpp]: Stray code.

**Comment by michaelweylandt** [comment]
I just took another pass through the changed C++ code - it's mostly type changes to Eigen containers, so easier to read than I anticipated. Looks pretty good throughout - the only place I got "surprised" was in the transpose for memory access - i might make that a bit clearer, but looks solid otherwise.

**Comment by jolars** [comment]
Great, yeah the transposing in the middle of the program is really not a clean solution and I would very much prefer something else. There are a few reasons for why I decided to do it anyway, though. First of all, most operations that operate across the columns of the original input (predictors and responses) can be performed early with the input in its original state. After this, almost every time-critical operation is performed sample-by-sample so transposing is obviously going to be effective.

My preferred solution, however, would be to keep everything in the same shape as it is given in the input, that is, samples stores in rows. However, that would necessitate that we use row-major storage in Eigen. This is possible but unfortunately the api isn't very welcoming to using .row() member functions and the developers themselves mention that support for row-major storage is not a priority. 

For now I'll document this better for the functions in the family class.

**Pull request merged**